modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsWriter.java

```diff
@@ -0,0 +1,517 @@
+        buf.putLong(uuid.getLeastSignificantBits());
+    }
+
+    /** Writes {@link IgniteUuid} to buffer. */
+    private static void writeIgniteUuid(ByteBuffer buf, IgniteUuid uuid) {
+        buf.putLong(uuid.globalId().getMostSignificantBits());
+        buf.putLong(uuid.globalId().getLeastSignificantBits());
+        buf.putLong(uuid.localId());
+    }
+
+    /** @return {@code True} if collecting performance statistics enabled. */
+    public boolean enabled() {
+        return enabled;
+    }
+
+    /** Stops collecting statistics in the cluster. */
+    void stopStatistics() {
+        try {
+            ctx.performanceStatistics().stopCollectStatistics();
+        }
+        catch (IgniteCheckedException e) {
+            log.error("Failed to stop performance statistics.", e);
+        }
+    }
+
+    /** Worker to write to performance statistics file. */
+    private class FileWriter extends GridWorker {
+        /**
+         * @param ctx Kernal context.
+         * @param log Logger.
+         */
+        FileWriter(GridKernalContext ctx, IgniteLogger log) {
+            super(ctx.igniteInstanceName(), "performance-statistics-writer%" + ctx.igniteInstanceName(), log,
+                ctx.workersRegistry());
+        }
+
+        /** {@inheritDoc} */
+        @Override protected void body() throws InterruptedException, IgniteInterruptedCheckedException {
+            while (!isCancelled()) {
+                blockingSectionBegin();
+
+                try {
+                    synchronized (this) {
+                        while (readyForFlushSize.get() < DFLT_FLUSH_SIZE && !isCancelled())
+                            wait();
+                    }
+                }
+                finally {
+                    blockingSectionEnd();
+                }
+
+                flushBuffer();
+            }
+
+            // Make sure that all producers released their buffers to safe deallocate memory.
+            flushBuffer();
+        }
+
+        /** Flushes to disk available bytes from the ring buffer. */
+        private void flushBuffer() {
+            List<SegmentedRingByteBuffer.ReadSegment> segs = ringByteBuffer.poll();
+
+            if (segs == null)
+                return;
+
+            try {
+                for (SegmentedRingByteBuffer.ReadSegment seg : segs) {
+                    updateHeartbeat();
+
+                    try {
+                        readyForFlushSize.addAndGet(-seg.buffer().remaining());
+
+                        fileIo.writeFully(seg.buffer());
+                    }
+                    finally {
+                        seg.release();
+                    }
+                }
+
+                fileIo.force();
+            } catch (IOException e) {
+                log.error("Unable to write to file. Performance statistics collecting will be stopped.", e);
+
+                stopStatistics();
+            }
+        }
+
+        /** {@inheritDoc} */
+        @Override public void cancel() {
```

Comment:
It seems we can just use regular `cancel` implementation if we will catch `InterruptedException` and handle it accordingly.

Reply:
Done.